client/components/image-preloader/index.tsx

```diff
@@ -0,0 +1,110 @@
+/**
+ * External dependencies
+ */
+import React, { useState, useEffect, useRef } from 'react';
+import { noop } from 'lodash';
+
+/**
+ * Constants
+ */
+const LoadStatus = {
+	PENDING: 'PENDING',
+	LOADING: 'LOADING',
+	LOADED: 'LOADED',
+	FAILED: 'FAILED',
+};
+
+type ImageEventHandler = ( event: string | Event ) => any;
+
+interface Props {
+	src?: string;
+	placeholder: React.ReactNode;
+	onLoad?: ImageEventHandler;
+	onError?: ImageEventHandler;
+}
+
+type ImgProps = Omit< React.ComponentProps< 'img' >, 'src' >;
+
+export default function ImagePreloader( props: Props & ImgProps ) {
+	const { children, src, placeholder, onLoad, onError, ...imageProps } = props;
+
+	const [ status, setStatus ] = useState( LoadStatus.PENDING );
```

Comment:
With the `enum` change above, I hope TypeScript infers this is `useState< LoadStatus >( … )`. If not, we should add the `LoadStatus` type variable.

Reply:
It does indeed!